### ISSUE TITLE
tests do not run when starting with flutter run -t

### ISSUE THREAD

**Issue by felixwortmann** [lulupointu/vrouter#46]
If you execute any test including a vrouter with `flutter run -t <path>`, the navigation does not work. It works when running without showing the UI (`flutter test`). I tested this by just running this code from `vwidget_test.dart`

```dart
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:vrouter/vrouter.dart';

main() {
  testWidgets('Stacked VWidget', (WidgetTester tester) async {
    await tester.pumpWidget(
      VRouter(
        routes: [
          VPage(
            path: '/',
            pageBuilder: (LocalKey key, Widget child) =>
                MaterialPage(key: key, child: child),
            widget: Builder(
              builder: (BuildContext context) => TextButton(
                child: Text('VWidget1'),
                onPressed: () => VRouter.of(context).push('/settings'),
              ),
            ),
            stackedRoutes: [
              VPage(
                path: '/settings',
                pageBuilder: (LocalKey key, Widget child) =>
                    MaterialPage(key: key, child: child),
                widget: Text('VWidget2'),
              ),
            ],
          ),
        ],
      ),
    );

    await tester.pumpAndSettle();

    // We should start at '/'

    final vWidget1Finder1 = find.text('VWidget1');
    final vWidget2Finder1 = find.text('VWidget2');

    expect(vWidget1Finder1, findsOneWidget);
    expect(vWidget2Finder1, findsNothing);

    // Try navigating to '/settings'
    // Tap the add button.
    await tester.tap(find.byType(TextButton));
    await tester.pumpAndSettle();

    // The navigation should have been redirected to / because we popped instead
    // So only VWidget should be visible
    final vWidget1Finder2 = find.text('VWidget1');
    final vWidget2Finder2 = find.text('VWidget2');

    expect(vWidget1Finder2, findsNothing);
    expect(vWidget2Finder2, findsOneWidget);
  });
}
```

This is the error:
```
flutter: ══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
flutter: The following TestFailure object was thrown running a test:
flutter:   Expected: no matching nodes in the widget tree
flutter:   Actual: _TextFinder:<exactly one widget with text "VWidget1" (ignoring offstage widgets):
flutter: Text("VWidget1", dependencies: [DefaultTextStyle, MediaQuery])>
flutter:    Which: means one was found but none were expected
```

**Comment by lulupointu** [comment]
`flutter run -t test/widget_test.dart` does not work for you?

I just tried and it is working for me:
* VRouter 1.1.0+20
* Flutter: Channel master, 2.2.0-11.0.pre.51
* Tested on chrome

What do you have ?

**Comment by felixwortmann** [comment]
Just tested in chrome, that works. <br>
I was testing on an IOS Simulator before, that still does not work.

**Comment by felixwortmann** [comment]
This is the output of `flutter doctor`
```
[✓] Flutter (Channel stable, 2.0.4, on macOS 11.2.3 20D91 darwin-x64, locale
    en-US)
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
[✓] Xcode - develop for iOS and macOS
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.1)
[✓] IntelliJ IDEA Ultimate Edition (version 2020.3.3)
[✓] VS Code (version 1.55.2)
[✓] Connected device (1 available)

• No issues found!
```

**Comment by lulupointu** [comment]
Oh well I'll try in IOS then. However you'll have to wait a week because I don't have my Mac with me right now, sorry about that

**Comment by felixwortmann** [comment]
No problem, since the tests are working fine without a device, this is not a pressing issue. I wanted to report it anyway.

**Comment by lulupointu** [comment]
Hi, I finally looked into this issue ! 

From what I understand, we were just not doing the right thing.
I am not an expert of integration testing since I've only used unit and widget testing in flutter so far but here is what I understand (correct me if I'm wrong).

Flutter has 3 types of testing:
* unit
* widget
* integration

The one we are interesting in is the 3rd one since we want to run the test on device. If I make an integration test and run it on device, it seems to work without issue.

<details>
<summary>
See this example:
</summary>

Your files should be:
```
test_app/
  lib/
  test_driver/
    app.dart
    app_test.dart
```

<details>
<summary>
app.dart:
</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter_driver/driver_extension.dart';
import 'package:vrouter/vrouter.dart';

void main() {
  enableFlutterDriverExtension();
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return VRouter(
      routes: [
        VPage(
          path: '/',
          pageBuilder: (LocalKey key, Widget child, name) =>
              MaterialPage(key: key, child: child, name: name),
          widget: Material(
            child: Builder(
              builder: (BuildContext context) => TextButton(
                child: Text('VWidget1'),
                onPressed: () => VRouter.of(context).push('/settings'),
              ),
            ),
          ),
          stackedRoutes: [
            VPage(
              path: '/settings',
              pageBuilder: (LocalKey key, Widget child, name) =>
                  MaterialPage(key: key, child: child, name: name),
              widget: Text('VWidget2'),
            ),
          ],
        ),
      ],
    );
  }
}
```

</details>

<details>
<summary>
app_test.dart:
</summary>

```dart
import 'package:flutter_driver/flutter_driver.dart';
import 'package:test/test.dart';

void main() {
  group('Stacked routes test', () {
    final vWidget1Finder = find.text('VWidget1');
    final vWidget2Finder = find.text('VWidget2');

    late FlutterDriver driver;

    // Connect to the Flutter driver before running any tests.
    setUpAll(() async {
      driver = await FlutterDriver.connect();
    });

    // Close the connection to the driver after the tests have completed.
    tearDownAll(() async {
      driver.close();
    });

    test("starts at '/'", () async {
      // Wait and find the text of the '/' page
      await driver.waitFor(vWidget1Finder);
    });

    test("Go to '/settings'", () async {
      // First, tap the button to navigate
      await driver.tap(vWidget1Finder);

      // Wait and find the text of the '/settings' page
      await driver.waitFor(vWidget2Finder);
    });
  });
}
```

</details>

You can run `flutter drive --target=test_driver/app.dart`with you emulator open, and you will see the test execute and succeed.

</details>

If we consider the first two. I don't think we are supposed to unit/widget test on device. 

<details>
<summary>
A "proof" of that is the following test, made **without** vrouter, which does not work with `flutter run`
</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:flutter_test/flutter_test.dart';

main() {
  testWidgets('Stacked VWidget', (WidgetTester tester) async {
    var count = 0;
    await tester.pumpWidget(
      MaterialApp(
        home: Material(
          child: Builder(
            builder: (BuildContext context) => TextButton(
              child: Text('VWidget1'),
              onPressed: () => count++,
            ),
          ),
        ),
      ),
    );

    await tester.pumpAndSettle();

    expect(count, 0);

    // Find the button
    final button = find.byType(InkWell);
    expect(button, findsOneWidget);

    // Tap the add button.
    await tester.tap(button);
    await tester.pumpAndSettle();

    expect(count, 1);
  });
}
```

If you run this test with `flutter run -t <path>` you will see that it fails. Even "better", if you look at the error message, it fails for the same reason **your** test (with vrouter) was failing. 

I think this proves that this is not the intended usage.

</details>

Now the true mystery is way **your** test (with vrouter) works on the web ?? I don't know but I'm not sure this is intended.

Please feel free to correct me if I've made a rookie mistake. As I said, I'm not an expert on integration testing.

**Comment by felixwortmann** [comment]
So, I looked back into this and you are right. I am trying to run a "widget test" on a device which is not the intended way for testing.<br> 
As it says [here](https://flutter.dev/docs/cookbook/testing/widget/finders) it is supported at least for debugging
>During a flutter run session on a widget test, you can also interactively tap parts of the screen for the Flutter tool to print the suggested [Finder].

I just tried again with a brand new flutter project and the default test. This also crashes on a device so the issue is not with vrouter. Maybe it is a flutter issue, it does not really matter because it is just for debugging tests and finders which it can. <br>Thanks for looking into this, closing it now.
